fix(defaults): visual star (*,#) is fragile

planet36 opened the pull request:
Close #21676

justinmk commented on runtime/doc/news.txt:
```diff
@@ -147,6 +147,11 @@ The following changes to existing APIs or features add new behavior.
+• The visual mode default-mappings ('*', '#') were changed to escape the
+  selected text.
+
+  See https://github.com/neovim/neovim/pull/21677.
```
Though the change is valuable, we generally should not add robustness improvements or most bug fixes in news.txt. It's not a changelog, it's a high-level newsletter.

You can ignore the failing "news.txt" CI job, or rename the PR/commit to `fix(..)` which is more appropriate anyway.

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -871,8 +871,8 @@ function vim._init_default_mappings()
-  map('x', '*', 'y/\\V<C-R>"<CR>')
-  map('x', '#', 'y?\\V<C-R>"<CR>')
+  map('x', '*', [[y/\V<C-R>=escape(@", '/\')<CR>]])
+  map('x', '#', [[y?\V<C-R>=escape(@", '?\')<CR>]])
```
I just noticed that these mappings are peforming a yank, wtf. 🤦 

Instead of another half-measure, may I request that we implement a full solution, as I linked to before: https://github.com/justinmk/config/blob/6e591bad58db8200e8eeebe20982ce3d1fc09c91/.config/nvim/init.vim#L794-L816

The current mappings also don't set `@/`, but they should.

planet36 replied:
Would it be better to yank to `/`?

justinmk replied:
that would be an improvement!

planet36 replied:
I was guessing out loud earlier, and after some trying I don't think it's possible to `"/y` like I thought.

I'll look at the vim script you linked and make something like it in lua.

ii14 replied:
Shouldn't `*` in visual mode also add work boundaries, `\<` and `\>`? vim-asterisk does this

planet36 replied:
> Shouldn't `*` in visual mode also add work boundaries, `\<` and `\>`? vim-asterisk does this

The characters after `\<` and before `\>` must be word characters.

planet36 replied:
> I just noticed that these mappings are peforming a yank, wtf. 🤦
> 
> Instead of another half-measure, may I request that we implement a full solution, as I linked to before: https://github.com/justinmk/config/blob/6e591bad58db8200e8eeebe20982ce3d1fc09c91/.config/nvim/init.vim#L794-L816

Correct me if I'm wrong, but this doesn't work when a multibyte character is the last one in a selection.  It only gets the first byte of the last character.

> 
> The current mappings also don't set `@/`, but they should.

The `/` register gets modified with the search command (right after the yank).

justinmk replied:
> Correct me if I'm wrong, but this doesn't work when a multibyte character is the last one in a selection. It only gets the first byte of the last character.

idk, improvements welcome.

georgejean replied:
Could something be built from
```vim
function! VisualSelection() abort
  try
    let unnamed_reg = @"
    let yank_reg = @0
    noautocmd silent! normal! gvy
    return @"
  finally
    let @" = unnamed_reg
    let @0 = yank_reg
  endtry
endfunction

xnoremap * <ESC>/\V<C-R>=substitute(escape(VisualSelection(), '/\'),'\n','\\n','g')<CR><CR>
xnoremap # <ESC>?\V<C-R>=substitute(escape(VisualSelection(), '?\'),'\n','\\n','g')<CR><CR>
```
(inspired from [#issuecomment-904514301](https://github.com/neovim/neovim/issues/15471#issuecomment-904514301) and [#issuecomment-1397813214](https://github.com/neovim/neovim/issues/21676#issuecomment-1397813214)) ?
It seems to work.

planet36 replied:
That performs a yank, which is what the maintainers want to avoid.

georgejean replied:
Yes, but here the different affected registers are saved and restored, which is not the case in version 0.9.0 of Neovim. I believe, unless I'm mistaken, that this is what can primarily bother the maintainers.

clason commented:
Please don't merge master into your PR branch; it messes up commit history. Instead rebase and force-push.

planet36 commented:
> Please don't merge master into your PR branch; it messes up commit history. Instead rebase and force-push.

Ok, sorry.  I was using the "Sync fork" button on github.

clason commented:
Yeah, that's a trap; it doesn't do what it seems to promise to do ;)

planet36 committed:
fix(defaults): visual mode star (*,#) is fragile

Problem:
Visual mode "*", "#" mappings don't work on text with "/", "\", "?", and
newlines.

Solution:
Get the visual selection and escape it as a search pattern.
Add functions vim.get_visual_selection and _search_for_visual_selection.

Fix #21676

planet36 commented:
This PR is ready to be discussed and reviewed.

justinmk commented on runtime/lua/vim/_editor.lua:
```diff
@@ -535,6 +535,55 @@ function vim.region(bufnr, pos1, pos2, regtype, inclusive)
+---        This will preserve the trailing newline of the selection when the result is concatenated with `"\n"`.
+---        Defaults to `false`.
+--- @return string|table
+function vim.get_visual_selection(list, append_empty)
```
Same comment that I made in https://github.com/neovim/neovim/pull/13896 and https://github.com/neovim/neovim/pull/21115#issuecomment-1322153511 

In particular #21115 has test coverage and re-uses low-level C internals which is likely a preferred starting point. 

Adding a new public function for _this_ PR can be avoided. I've linked to https://github.com/justinmk/config/blob/6e591bad58db8200e8eeebe20982ce3d1fc09c91/.config/nvim/init.vim#L794-L816 several times, why not use it? It's less code.

planet36 replied:
Thanks for looking at this.

I first tried using `vim.region` but it was challenging when the last character of a selection was multibyte.  IIRC `vim.region` returns the byte index of the last character selected, but I need to know how many bytes are in that character (to add to the `end_col` param in `nvim_buf_get_text`).

I called `nvim_buf_get_lines` to get the whole line, but I'd still need to find how many bytes are in the last character of the selection.
Given the whole line, I tried using `strgetchar`, `nr2char`, `byteidx`, and `string.sub` in various ways, but it got complicated.

I eventually gave up and used the yank approach because it just works.

Some of my pain points when using the `vim.region`, `nvim_buf_get_lines`, `nvim_buf_get_text` approach:
* `nvim_buf_get_text` does't have a `strict_indexing` option like `nvim_buf_get_lines` has.
* Lua 5.1 isn't Unicode-aware
* The APIs take/return 0-index values, but Lua arrays, row numbers, and column numbers start at 1.  (I know I should "git gud", but I was always second-guessing if I needed to add or subtract 1)

justinmk replied:
Those are valid problems, but let's solve them instead of adding a new api.

> nvim_buf_get_text doesn't have a strict_indexing option like nvim_buf_get_lines has. 

Could add a `opts.strict_indexing` flag to nvim_buf_get_text.

> Lua 5.1 isn't Unicode-aware

There is no such thing really ([see also](https://github.com/neovim/neovim/issues/68#issuecomment-170268097)). But `strgetchar`, `strchars`, `strwidth` are available from Lua.

justinmk commented on runtime/doc/vim_diff.txt:
```diff
@@ -119,10 +119,11 @@ of these in your config by simply removing the mapping, e.g. ":unmap Y".
-	xnoremap # y?\V<C-R>"<CR>
 	nnoremap & :&&<CR>
 <
+Default mappings composed of Lua code are not listed above. Use ":map" to see
+the |map-listing|.
```
For cases where listing the exact mapping is too much, we can link to the defaults explicitly here. It's useful to have a single place where all the default mappings are linked from.

```
|v_#-default|
|v_star-default|
```

justinmk committed:
refactor(defaults): use vim.region for visual star (*,#)

Problem:
The parent commit added a new vim.get_visual_selection() function to
improve visual star. But that is redundant with vim.region(). Any
current limitations of vim.region() should be fixed instead of adding
a new function.

Solution:
Delete vim.get_visual_selection().
Use vim.region() to get the visual selection.
TODO: fails with visual "block" selections.

justinmk commented:
Pushed an update that uses `vim.region()` instead of adding a new function.

clason commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1010,18 +1010,54 @@ end
+  ---@private
   local function map(mode, lhs, rhs)
-    vim.api.nvim_set_keymap(mode, lhs, rhs, { noremap = true, desc = 'Nvim builtin' })
+    vim.keymap.set(mode, lhs, rhs, { noremap = true, desc = 'Nvim builtin' })
```
```suggestion
    vim.keymap.set(mode, lhs, rhs, { desc = 'Nvim builtin' })
```
`vim.keymap.set` does not support `noremap` (and `remap=false` is the default anyway)

adelarsq commented:
@justinmk So with these changes also add an easy way to get current selected text? `get_visual_selection` will be available outside?

justinmk commented:
> @justinmk So with these changes also add an easy way to get current selected text? `get_visual_selection` will be available outside?

That isn't in the scope of this PR. I explained in https://github.com/neovim/neovim/pull/21677#discussion_r1252991079 that `vim.region` should be enhanced to better support this, i.e. we should fix its limitations. See that comment for details.

Currently to get the visual selection you can use the approach shown in this PR:

```lua
local function region_chunks(region)
  local chunks = {}
  local maxcol = vim.v.maxcol
  for line, cols in vim.spairs(region) do
    local endcol = cols[2] == maxcol and -1 or cols[2]
    local chunk = vim.api.nvim_buf_get_text(0, line, cols[1], line, endcol, {})[1]
    table.insert(chunks, chunk)
  end
  return chunks
end


local region = vim.region(0, "'<", "'>", vim.fn.visualmode(), vim.o.selection == 'inclusive')
local chunks = region_chunks(region)
```

However that has the multibyte limitations discussed in the linked comment. Also doesn't work with visual block selections (this is a `vim.region` bug, should be fixable).

planet36 commented:
> @justinmk So with these changes also add an easy way to get current selected text? `get_visual_selection` will be available outside?

You can add it to your init.lua if it suits your needs.

echasnovski commented on runtime/lua/vim/_editor.lua:
```diff
@@ -1000,28 +1000,59 @@ function vim.deprecate(name, alternative, version, plugin, backtrace)
+    local esc_chunks = vim
+      .iter(chunks)
+      :map(function(v)
+        return vim.fn.escape(v, [[/\]])
```
@planet36, when `cmd` is `?`, escaped characters originally were `[[?\]]`. Is there a reason to dropping this?

With current approach, `#` fails on the following example after `v$h#` with cursor on the start of the first line:

```
aa?/\bb
aa
aa?/\bb
```

justinmk replied:
That's my mistake, thanks for mentioning it
